### DISCUSSION TITLE
Zeitplan für Stufenschalter - nur mit Überschussenergie = min. Energie 0?

### DISCUSSION THREAD

**Discussion by andi5783** [camueller/SmartApplianceEnabler#481] [Allgemein]
Hallo zusammen,
eine Frage bzgl. der Zeitpläne für Stufenschalter, Anforderungsart Energie:

- Gerät: Wärmepumpe - min. Energie 1500W max. Energie 6500W (in Wirklichkeit ioBroker der über Modbus Aktionen in der Wärmepumpe auslöst)
- Stufenschalter mit 2 HTTP-Schalter: 2 Stufen, die erste hat 1500W und die zweite 6500W
- Im Sunny Portal: Sehr hohe Priorität, 95% aus PV 

Zeitpläne:
1. 6-12 Uhr min. Energie = 0Wh, max. Energie 20000Wh
2. 12-14 Uhr min. Energie = 2000Wh, max. Energie 13000Wh
3. 14-20 Uhr min. Energie = 0 Wh, max. Energie 20000Wh

Idee:
- Im Zeitfenster 2 soll zumindest die Stufe 1 für min. ca. 1h ausgelöst werden (z.B. WW-Bereitung auf 57 Grad) es darf aber auch gerne die Stufe 2 (z.B. WW-Bereitung auf 65 Grad) sein, falls genügend Überschuss vorhanden ist.
- Im Zeitfenster 1 und 3 sollen die Stufen nur ausgelöst werden werden wenn Überschuss vorhanden ist. 

Ergebnis:
Der 1. und 3. Zeitplan wurden entsprechend im Sunny Portal durchsichtig angezeigt, für den zweiten Plan hat er entsprechend einen undurchsichtigen Balken angezeigt, weil ja min. Ladung gefordert. Soweit so gut, aber:

Die Schalter selbst wurden nur für den 2.Zeitplan geschaltet, d.h. der SHM hat dem SAE kurz nach 12 mitgeteilt, dass Gerät einzuschalten und zwar in der 6500W Stufe. Um 14 Uhr wurde es entsprechend abgeschaltet.

Im 1. und 3. ist nichts passiert, obwohl eigentlich genügend Leistung verfügbar war. Anhand der Doku denke ich die Ursache zu haben, denn dort steht 

"_Normalerweise wird nur die **max. Energie** angegeben, die **auf jeden Fall geladen werden soll**.

Optional kann für die min. Energie ein kleinerer Wert angegeben werden. Falls er angegeben ist, wird nur dieser Wert auf jeden Fall bereitgestellt und die darüber hinausgehende Energiemenge bis zur max. Ernergie nur dann, wenn Überschussenergie verfügbar ist._"

Heißt das, dass bei Angabe von min. Energie = 0 wirklich die max. Energie geladen werden muss, was dann natürlich nicht geklappt hätte? Oder umgekehrt beschrieben, die Angabe von min. Energie = 0 heißt eben nicht, dass nur Überschussenergie verwendet werden soll bis zur max. Energie, sondern das die max. Energie?

Ich werde jetzt mal als min. Energie 1Wh angeben um meine Theorie zu testen, würde mich aber über euer Feedback freuen.

Schöne Grüße
Andi

**Comment by andi5783** [comment]
Mir ist gerade bewusst geworden, dass die 3 Zeitpläne an sich Unsinn sind. Ich könnte nur einen machen, mit min. Energie 2000Wh und dann sucht sich der SHM das beste Zeitfenster raus.

Trotzdem wäre noch interessant ob min. Energie = 0 reines Überschussladen bedeutet oder das max. Energie geladen werden muss.

Danke schon mal!

**Comment by camueller** [comment]
> Im 1. und 3. ist nichts passiert, obwohl eigentlich genügend Leistung verfügbar war.

Um es dem Sunny Home Manager einfacher zu machen, solltest Du keine `min. Einschaltdauer` im SAE beim dem Gerät angeben. Ausserdem sollte `Unterbrechung erlaubt` ausgewählt sein, auch wenn das für die Wärempumpe eigentlich nicht gut ist.
Im Sunny Portal solltest Du prüfen, dass die Anforderungen für Netzbezug nicht zu streng sind für das Gerät.

Wenn Du bei genügen Überschuss Einschaltbefehle vom Sunny Home Manager bekommst, kannst Du die genannten Bedingungen wieder verschärfen.

> Optional kann für die min. Energie ein kleinerer Wert angegeben werden. Falls er angegeben ist, wird nur dieser Wert auf jeden Fall bereitgestellt und die darüber hinausgehende Energiemenge bis zur max. Ernergie nur dann, wenn Überschussenergie verfügbar ist._"

> Heißt das, dass bei Angabe von min. Energie = 0 wirklich die max. Energie geladen werden muss, was dann natürlich nicht geklappt hätte? Oder umgekehrt beschrieben, die Angabe von min. Energie = 0 heißt eben nicht, dass nur Überschussenergie verwendet werden soll bis zur max. Energie, sondern das die max. Energie?

Eigentlich ist der Satz in der Dokumentation eindeutig formuliert. Insofern verstehe ich nciht, warum Du ihn so interpretierst, dass bei "min. Energie = 0 wirklich die max. Energie geladen werden muss". Wenn man 0 als min. Energie angibt, muss nur dieser Wert (also 0) bereitgestellt werden.

**Reply by andi5783** [reply]
Hi und vielen Dank für die Tipps, werde ich entsprechend ausprobieren. 

Die Interpretation kam daher, weil ich nach möglichen Ursachen gesucht habe. Vom Text her ist es eindeutig und es ist Ja auch logisch, weil es dann so umgesetzt ist wie bei der Anforderungsart Laufzeit. Auf Grund der Tatsache, dass es sich um einen variablen Verbraucher handelt, hab ich dann aber doch in Erwägung gezogen, dass SMA/SAE hier anders tickt. 

Besten Dank nochmals für die Antwort 👍🏼

**Comment by camueller** [comment]
Mir ist noch etwas eingefallen: Hast Du im SAE für das Gerät ein min. Leistungsaufnahme konfiguriert? Wenn die fehlt, wird das Gerät vom SHM nicht als Verbraucher mit variabler Leistungsaufnahme erkannt.

**Reply by andi5783** [reply]
Ja, die habe ich angegeben. Hab jetzt mal die min. Einschaltdauer rausgenommen, Netzbezug auf 10% gestellt und SHM hat direkt umgeplant. Heute wird es aber generell schwierig, da die Prognose eher gering ist.

**Reply by andi5783** [reply]
Also mit den Anpassungen funktioniert es heute einwandfrei und er schaltet die Stufen wenn Energie verfügbar.

**Comment by camueller** [comment]
Super - Danke für das Feedback!

**Comment by camueller** [comment]
Welche Informationen haben in der Dokumentation gefehlt oder waren unklar, sodass Du nicht direkt erfolgreich warst mit dem Stufenschalter? Dieser Schaltertyp ist noch recht neu ...

Der folgende Link müsste für Dich auch interessant sein, weil auch der Stufenschalter bei Verbrauchern mit variabler Leistungsaufnahme zum Einsatz kommt:
https://github.com/camueller/SmartApplianceEnabler/discussions/285#discussioncomment-5065297

**Reply by andi5783** [reply]
Also ein Thema war, dass mir nicht klar war, dass min./max. Einschalt-/Ausschaltdauer es recht kompliziert machen für den SHM. Wenn der Hinweis drinnen wäre, dann wäre das schon mal hilfreich.

In der Schedule-Beschreibung würde ich den Text 

"_Mit der Anforderungsart Energie kann wird anzufordernde Energie direkt vorgegeben. Normalerweise wird nur die max. Energie angegeben, die auf jeden Fall geladen werden soll. Optional kann für die min. Energie ein kleinerer Wert angegeben werden. Falls er angegeben ist, wird nur dieser Wert auf jeden Fall bereitgestellt und die darüber hinausgehende Energiemenge bis zur max. Ernergie nur dann, wenn Überschussenergie verfügbar ist_"

durch folgenden ersetzen, damit es analog zur Zeitanforderung ist:

"_Mit der Anforderungsart Energie kann wird anzufordernde Energie direkt vorgegeben. Dabei wird über die Startzeit und die Endzeit ein Zeitfenster definiert, innerhalb dessen der Sunny Home Manager die maximale Energie unabhängig vom Vorhandensein von PV-Strom sicherstellen muss. Wird zusätzlich die (optionale) minimale Energie angegeben, wird der Sunny Home Manager nur diese Energie sicher bereitstellen, aber bei Vorhandensein von Überschussenenergie die bereitgestellte Energie bis zur maximalen Energie erweitern. Das Setzen einer minimalen Energie von 0 führt dazu, dass das Gerät ausschliesslich mit Überschussenergie betrieben wird. Wenn diese nicht vorhanden ist, wird das Gerät nicht eingeschaltet_"

Dann kann man das mit der min. Energie = 0 direkt rauslesen. Zumindest mich hat der andere Text etwas verunsichert.

Ich hatte heute dann doch noch einen größeren Crash mit dem SAE:
Mir ist nach 16 Uhr aufgefallen, dass meine Batterie (11kw) auf 7kw entladen wurde und noch am entladen ist. Ursache war hoher Verbrauch und ein geringer PV Ertrag. Als ich mir das Ganze nach der Arbeit gegen 16:45 Uhr genauer anschauen wollte, ist mir aufgefallen dass der Schalter den der SAE ansteuert ständig on/off springt. Auf dem ioBroker, ist mir aufgefallen das massive Perfomanceprobleme vorliegen, im dortigen Log ist mir dann aufgefallen, dass das Log von MQTT Messages geflutet wird (hatte # abonniert). Ursache schein ein Amoklauf des SAE oder SHM gewesen zu sein (konnte ich noch nicht herausarbeiten), was unzählige MQTT Messages pro Sekunde/Millisekunde verursacht hat. Einzige Lösung, war ein Shutdown von SAE und dem MQTT Broker. Nach dem Neustart habe ich jetzt erstmal den Zeitplan deaktiviert und hatte keine Auffälligkeiten mehr.

Evtl. kannst du mal einen Blick ins Log werfen, ob du eine Ursache ausmachst. Hab es [hier](https://1drv.ms/u/s!AkqO8z9FwMwZj-RNSjdBjI0v2m60tg?e=RppYIm) abgelegt. Bin noch nicht fit und schaue es mir momentan noch via Texteditor an und weiß nicht was ich suchen muss.

**Reply by andi5783** [reply]
Gerade ist mir aufgefallen, dass er doch noch Amok läuft. Hab den SAE jetzt mal abgeschaltet. Irgendwas passt nicht mim SAE oder SHM.

[Hier das zweite, jetzt sehr große Log](https://1drv.ms/u/s!AkqO8z9FwMwZj-ROCbQPKFdPiGkROw?e=vH6nMg)....das ging anscheinend eine ganze weile...

**Comment by camueller** [comment]
Deinen Vorschlag für die Änderung der Dokumentation habe ich weitgehend übernommen. Danke dafür!

Zu Deinem Log-File. Das zeigt 6 Timertasks für den Stufenschalter des Gerätes `100000000005`:
```
2023-02-23 13:25:02,874 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=1461483305
2023-02-23 13:25:02,917 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=63918295
2023-02-23 13:25:03,145 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=1104247047
2023-02-23 13:25:03,162 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=201209055
2023-02-23 13:25:03,211 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=367606045
2023-02-23 13:25:03,298 TRACE [Timer-0] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:54] F-00000001-100000000005-00: Executing timer task name=MqttPublish-LevelSwitch id=778199724
```
Normalerweise darf es nur einen TimerTask geben. Ich untersuche das mal ...

**Comment by camueller** [comment]
Kann es sein, dass Du seit dem letzten SAE-Start immer wieder die Konfiguration geändert hast? Dabei wurden offensichtlich diese Tasks nicht gestoppt und bei jeder Konfigurationsänderung kam einer dazu. Wenn Du den SAE neu startest, sollte das Problem weg sein. Natürlich fixe ich das auch im Code, damit es künftig nicht mehr auftritt.

**Reply by andi5783** [reply]
Also deine Vermutung trifft mit Sicherheit für das erste Auftreten zu, da ich noch einige Anpassungen gemacht habe ohne den SAE neu zustarten, ABER ein Neustart hat hier nicht geholfen um das Problem zu lösen:
Nach dem initialen Erkennen des Problems, habe ich den SAE und den MQTT gestoppt und anschließend beide wieder gestartet. Im SAE hatte ich dann direkt den Zeitplan deaktiviert. Das Problem ist aber wieder aufgetreten und ich habe es dann kurz vor 23 Uhr gemerkt. Also hab ich ihn wieder gestoppt. Anschließend habe ihn wieder gestartet und heute in der Früh hatte ich wieder das Problem. Es ist fast so, als wären alte Einträge irgendwo gespeichert.

Und wo ich das schreibe, kommt mir noch ein Hinweis. Ich hatte schon ähnliche Probleme:
Ich hatte ursprünglich schon andere Geräte gespeichert/testweise angelegt. Nach dem ich die ersten Gerätevariante betriebsbereit hatte, ist mir aufgefallen, dass in den MQTT Meldungen in den SAE Logs noch Geräte auftauchen, die schon längst gelöscht waren. Selbst ein Neustart des SAE hat hier nicht geholfen und die Geräte sind weiter fröhlich in den Logs erschienen. Erst als ich den persistenten Speicher von MQTT abgeschaltet und alles gelöscht hatte, waren die Geräte verschwunden. Mittlerweile werden die MQTT Daten nicht mehr persistent gespeichert, dass kann also theoretisch nicht die Ursache für die TimerTasks sein...aber ggf. hat hier die MQTT Config und damit die nicht-persistente Speicherung noch nicht gewirkt. Den ggf. hätte ich die Container Station im QNAP Neustarten müssen, damit die Config wirkt. Heute Nacht wurde die NAS auf Grund eines Update neugestartet, daher sollte die persistente Speicherung definitiv weg sein....
Aber trotzdem als Gedankenanstoss: Liest SAE irgendetwas aus dem MQTT Broker aus, wenn dort noch alte Daten sind?

**Comment by camueller** [comment]
Ich habe eine Test-Version mit dem Fix für Dich gebaut: https://drive.google.com/file/d/1NTTmyNCaiKtpQsGADBOR1QQ5e5KtEvIn/view?usp=share_link

Es wäre schön, wenn Du diese zeitnah testen könntest, da ich bald ein offizielles Bugfix-Release bauen will.

**Reply by andi5783** [reply]
Lade ich direkt runter und spiele ich gleich noch ein. Morgen schau ich dann wie es läuft. Kann dir aber evtl. erst morgen Abend bescheid geben, da ich mit der Familie in Hamburg auf einem Musical bin.

**Comment by camueller** [comment]
> Nach dem initialen Erkennen des Problems, habe ich den SAE und den MQTT gestoppt und anschließend beide wieder gestartet. Im SAE hatte ich dann direkt den Zeitplan deaktiviert.

Jede Änderung der Konfiguration inkl. Zeitpläne bewirkt eine Neu-Initialisierung des SAE, welche zu den mehrfach vorhandenen TimerTask geführt hat. Dadurch wurden MQTT-Nachrichten mehrfach verarbeitet. Das habe ich in der gestern gebauten Test-Version behoben.

> Aber trotzdem als Gedankenanstoss: Liest SAE irgendetwas aus dem MQTT Broker aus, wenn dort noch alte Daten sind?

Die einzigen Nachrichten, welche vom MQTT-Broker aufbewahrt ("retained") werden, sind die mit der Timeframe-Queue, weil diese nur beim SAE-Starte, bei Änderungen und beim Neufüllen der Queue neu gesendet wird. Das macht aber normalerweise kein Problem.

**Reply by andi5783** [reply]
Passt, verstanden 👍🏼

Mit der neuen Version läuft bis jetzt alles nach Plan. Zuerst Stufe 1 und aktuell hat er wegen Reserven Stufe 2 ausgelöst. 

Ich werde dir heute Abend mal das Log schicken. Evtl. erkennst du noch Probleme die ich so nicht merke.

**Comment by andi5783** [comment]
Hi, das Problem ist wieder aufgetreten. Hier die Log-Files vom [25.2.](https://www.myqnapcloud.com/smartshare/692j460k7p84o211rw883by0_3gg9ef4kl1012190q1ty1wy8dz8beg09) und [26.2.](https://www.myqnapcloud.com/smartshare/692j460k7p84o211rw883by0_34cfh3ik36ln2q8sqruuy36b8ceb2g1j) (das Problem ist am 26.2. gegen 5:14 Uhr aufgetreten und daher hat das Log-File knappe 2GB)

Kurzer Ablauf was ich gemacht habe bzw. was passiert ist:
- Tagsüber ist der SAE normal gelaufen und der SHM hat entsprechende Schaltbefehle gesendet. 
- Mir ist dann am späten Nachmittag aufgefallen, dass wieder mein Akku geleert wird und gleichzeitig 2500W extern bezogen werde, da keine PV Leistung vorhanden ist, aber der Stufenschalter auf der 2.Stufe d.h. 6500W läuft. Ich habe im Hintergrund den ioBroker daraufhin so eingestellt, dass er die Schaltbefehle empfängt, aber nicht umsetzt. Dadurch hat man die Logs was der SAE noch gemacht hat ohne dass mein Akku geleert wird.
- Nachts habe ich dann die Logs angeschaut. Es kamen zwei Schaltbefehle die ich nicht nachvollziehen kann und zwar um 15:05 Uhr und 16:21 Uhr. Hier wurden 6500 angefordert, obwohl nicht ansatzweise Energie vorhanden war (PV 0,5KW). Im SHM ist ein Verhältnis mit 10% Netzbezug eingestellt. 

Mögliche Ursache, die ich mir vorstellen kann:
Es liegt daran, dass ich aktuell noch keinen Zähler für das Gerät habe und der SHM versucht die angegeben min. 3KW zu laden.

Zur Endzeit (bei mir 20 Uhr) kam dann um 19:59 Uhr der Abschaltbefehl.

- ich habe 0KW min. Energie angegeben um dann das Verhalten am nächsten Tag zu prüfen und den SAE neugestartet. Dabei gab es kein Auffälligkeiten
- Sonntag Nachmittag habe ich festgestellt, dass der SAE wieder Amok läuft.
- Ich habe ihn abgestellt und die Logs geprüft:
Um 5Uhr sonntags wird mein QNAP wöchentlich neugestartet. Der SAE ist auch entsprechend hochgefahren und ist dann ab 5:14 Uhr Amok gelaufen. D.h. der Neustart wenige Stunden vorher war kein Problem, sondern erst der zweite um 5:14 Uhr.

Fällt dir noch was auf? Teste gerne eine weitere Version.

**Comment by camueller** [comment]
Irgendetwas stimmt mit Deinem MQTT-Setup nicht. Um 05:00:55 ging die Verbindung verloren:
```
2023-02-26 05:00:55,797 ERROR [MQTT Rec: ApplianceManager-0] d.a.s.m.MqttClient [MqttClient.java:166] ApplianceManager: MQTT connecton lost
```

Um 05:13:39 hast Du den SAE neu gestartet - wozu soll das gut sein?
```
2023-02-26 05:13:39,585 INFO [main] d.a.s.Application [StartupInfoLogger.java:55] Starting Application v2.2.1 on sae with PID 7 (/opt/sae/SmartApplianceEnabler-2.2.1.war started by root in /)
```

Der neu gestartete SAE kann einen der HTTP-Schalter nicht erreichen:
```
2023-02-26 05:14:36,113 ERROR [MQTT Call: F-05071983-100000000005-00-HttpSwitch-1] d.a.s.h.HttpTransactionExecutor [HttpTransactionExecutor.java:114] F-05071983-100000000005-00: Error executing GET request.
org.apache.http.conn.HttpHostConnectException: Connect to 192.168.10.7:8087 [/192.168.10.7] failed: Connection refused (Connection refused)
```

An der Stelle breche ich die Analyse ab - das ist mir zuviel Chaos. Wenn Du den SAE normal betreibst (ohne Restarts oder nicht erreichbar MQTT-Broker bzw. Schalter) schaue ich mir gern wieder ein Log an.

**Reply by andi5783** [reply]
Das ist schade, ich betreibe das ganze vollkommen „normal“. Das einzige das um 5 Uhr neugestartet wurde, war das QNAP NAS. Auf diesem Laufen virtualisiert: SAE, MQTT und der ioBroker (der den http Schalter beinhaltet). 
Daher sind alle diese Komponenten neugestartet worden, was die von dir erwähnten Meldungen erklärt. 

Mit dem eigentlichen Problem sollte das nichts zu tun haben, denn beim letzten Mal gab es keinen Neustart. 

Ich lasse es einfach mal weiterlaufen, ich gehe davon aus, dass der Fehler wieder auftritt und dann ist kein Neustart dabei (dieser erfolgt bloß Sonntags).

**Reply by camueller** [reply]
Wozu soll es gut sein irgendetwas regelmäßig neu zu starten? Das mag für Windows sinnvoll sein, aber unter Linux muss man nichts periodisch neu starten.

Du solltest den SAE  jetzt aber nochmal neu starten, nachdem MQTT und ioBroker erreichbar sind. Sonst ist das kein sauberer Ausgangszustand für diesen Test, da möglicherweise durch das Restart-Chaos die Ausgangsbedingungen andere waren.

**Reply by andi5783** [reply]
Puh, das weiß ich eigentlich gar nicht mehr genau 🙈 

Ich glaube, als ich mir das NAS vor 6 Jahren angeschafft hab, hatte ich Probleme, dass es ab und an nicht erreichbar war….oder war es der Mediastreaming-Dienst der darauf lief…der wöchentliche Neustart hat damals geholfen. War halt die einfachste/schnellste Lösung, die ich als alter Windows/.Net-Programmierer gewählt hab 😄

Da er mich mittlerweile aber auch an anderen Stellen stört, hab ich das vorhin ausgeschaltet 👍🏼

**Comment by camueller** [comment]
Kleine Info am Rande: Ich bin ab Dienstag ein paar Tage nur eingeschränkt erreichbar (Krankenhaus/OP) und habe dort nur mein Handy, aber keinen Laptop. In der Zeit kannst Du schauen, ob der Fehler nochmal auftritt :-)

**Reply by andi5783** [reply]
Ja, ich werde das genau beobachten und mich auch noch mit Linux und dem Log besser vertraut machen. Dir alles Gute für die OP und eine schnelle Genesung!

**Comment by andi5783** [comment]
So, ich hoffe du hast soweit alles gut überstanden!

Ich hab gestern nochmal meine Geräte neu angelegt und heute war quasi der erste volle Tag:
Alles hat normal ausgeschaut und die Stufen wurden aktiviert. Zwischen 12 und13 Uhr bin ich dann auf die Idee gekommen meine Lüftungsanlage hinzuzufügen (Gerät + HTTP-Zähler + Zählerbasierter Zustandsmelder) um den Verbrauch auch im Sunny Portal sichtbar zu haben. Nach dem hinzufügen hatte ich dann den ersten Amoklauf, was die MQTT Nachrichten angeht und ich musst den SAE neustarten. Danach lief er normal weiter.

Später Nachmittag, kurz nach 16 Uhr:
Obwohl die PV Leistung bereits zu schwach ist, sendet der SHM die Empfehlung für die 2. Stufe mit 5500kw, das heißt die Batterie wird verbraucht (dazu mache  ich mal eine separate Diskussion, damit es nicht total chaotisch wird). Die einzige Erklärung die ich dafür hatte war, dass die eingestellte min. Energie dies auslöst, daher habe ich die min. Energie auf 0wh gesetzt und gespeichert. Mit dem Speichern kam dann der zweite Amoklauf und ich musste den SAE wieder Neustarten. Danach lief alles wieder normal und der SHM hat dann auch ein Request "Received control request: on=false" gesendet.

Es scheint mir und den Eindruck hatte ich vorher schon an der ein oder anderen Stelle, dass er irgendwie Probleme beim Speichern hat.

[Hier](https://www.myqnapcloud.com/smartshare/692j460k7p84o211rw883by0_74f4i4h00m5020tt04s99w8b42c6f5j9) wäre die Log-Datei, falls du mal Zeit zum draufschauen hast.

Vielen Dank schon mal!

**Comment by camueller** [comment]
Ich habe mir das Log mal angesehen: Um 12:38:40 hast Du die Lüftung als Gerät angelegt:
```
2023-03-01 12:38:40,662 DEBUG [http-nio-8080-exec-9] d.a.s.w.SaeController [SaeController.java:198] F-05071983-050719830003-00: Received request to set ApplianceInfo (create=true): ApplianceInfo{id='F-05071983-050719830003-00', name='Lüftung', serial='050719830003', type=Other, vendor='LTM', minPowerConsumption=null, maxPowerConsumption=245, minOnTime=null, maxOnTime=null, minOffTime=null, maxOffTime=null, currentPowerMethod=null, interruptionsAllowed=false, notificationSenderId=null}
```
Dabei wurden die TimerTasks korrekt gestoppt - das war ja mein Fix in der Test-Version, die ich für Dich gebaut hatte:
```
2023-03-01 12:38:41,627 DEBUG [http-nio-8080-exec-9] d.a.s.c.LevelSwitch [LevelSwitch.java:194] F-05071983-050719830002-00: Stopping ...
2023-03-01 12:38:41,627 TRACE [http-nio-8080-exec-9] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:66] F-05071983-050719830002-00: Cancel timer task name=MqttPublish-HttpSwitch id=919731753
2023-03-01 12:38:41,670 DEBUG [http-nio-8080-exec-9] d.a.s.m.MqttClient [MqttClient.java:203] F-05071983-050719830002-00-MQTT-HttpSwitch: disconnected from MQTT broker - was connected=true
2023-03-01 12:38:41,671 TRACE [http-nio-8080-exec-9] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:66] F-05071983-050719830002-00: Cancel timer task name=MqttPublish-HttpSwitch id=1234673049
2023-03-01 12:38:41,697 DEBUG [http-nio-8080-exec-9] d.a.s.m.MqttClient [MqttClient.java:203] F-05071983-050719830002-00-MQTT-HttpSwitch: disconnected from MQTT broker - was connected=true
2023-03-01 12:38:41,711 TRACE [http-nio-8080-exec-9] d.a.s.u.GuardedTimerTask [GuardedTimerTask.java:66] F-05071983-050719830002-00: Cancel timer task name=MqttPublish-LevelSwitch id=1564580572
2023-03-01 12:38:41,740 DEBUG [http-nio-8080-exec-9] d.a.s.m.MqttClient [MqttClient.java:203] F-05071983-050719830002-00-MQTT-LevelSwitch: disconnected from MQTT broker - was connected=true
```
Während der Neu-Initialisierung wird plötzlich eine MQTT-Nachricht empfangen, dass das Gerät wieder mit 5500W eingeschaltet werden soll.
```
2023-03-01 12:38:46,186 TRACE [MQTT Call: F-05071983-050719830002-00-LevelSwitch-2] d.a.s.m.MqttClient [MqttClient.java:318] F-05071983-050719830002-00-MQTT-LevelSwitch: Message received: topic=sae/F-05071983-050719830002-00/Control/set payload={"on":true,"power":5500,"time":"2023-03-01T12:05:06.41241","type":"VariablePowerConsumerMessage","useOptionalEnergy":null}
```
Diese Nachricht hat der SAE aber gar nicht nach dem Hinzufügen der Lüftung gesendet (falls sie überhaupt vom SAE erzeugt wurde). Sie wurde um bereits 12:05:06 erzeugt und hätte vom SAE zu diesem Zeitpunkt nicht empfangen werden dürfen. Da stimmt etwas mit Deinem MQTT-Broker nicht.

**Reply by andi5783** [reply]
Ich glaube wir kommen der Sache näher:
Die erwähnte Nachricht wurde vom SAE erzeugt und zwar 👍 
`2023-03-01 12:05:06,414 TRACE [MQTT Call: F-05071983-050719830002-00-LevelSwitch-1] d.a.s.m.MqttClient [MqttClient.java:318] F-05071983-050719830002-00-MQTT-LevelSwitch: Message received: topic=sae/F-05071983-050719830002-00/Control/set payload={"on":true,"power":5500,"time":"2023-03-01T12:05:06.41241","type":"VariablePowerConsumerMessage","useOptionalEnergy":null}`

Kann es damit zusammenhängen das du für die MQTT Nachrichten das "Retain Flag" setzt und sich der SAE deswegen beim Neu-Initialisieren die alten Nachrichten holt, die er vorher gepublished hat?

**Comment by camueller** [comment]
> Kann es damit zusammenhängen das du für die MQTT Nachrichten das "Retain Flag" setzt und sich der SAE deswegen beim Neu-Initialisieren die alten Nachrichten holt, die er vorher gepublished hat?

Kurz Antwort: Ja.

Anders, als ich weiter oben geschrieben hatte, hatte ich für diverse MQTT-Nachrichten das Retained-Flag gesetzt. Als ich mich dafür entschieden hatte, habe ich vermutlich nicht die Konsequenzen beim Neu-Initialisieren bedacht, wenn "alte" MQTT-Nachrichten zugestellt werden. Eigentlich gibt es keine echte Notwendigkeit für das Retained-Flag im Kontext des SAE. Deshalb habe ich es testweise an allen Stellen deaktiviert. Ausserdem wird jetzt beim Versenden der MQTT-Nachrichten auch gelogt, ob das Retained-Flag gesetzt wird.

Hier eine entsprechende Test-Version:
https://drive.google.com/file/d/1arvpov56Wt7xaTIFkSMg5CYF0P0m9CG4/view?usp=share_link

**Reply by andi5783** [reply]
Perfekt, super. Spiele ich heute Abend ein und teste es entsprechend.

**Comment by andi5783** [comment]
So, hab das ganze eingespielt und direkt ein wenig dann getestet in dem ich Einstellungen gespeichert habe. Bislang keine Probleme mehr.

Morgen teste ich es auch nochmal.

**Comment by andi5783** [comment]
Hallo zusammen,
ich konnte mit der Testversion keine Probleme mehr mit MQTT feststellen.

Besten Dank für den Fix :)

**Comment by camueller** [comment]
Danke für Dein Feedback! Davon werden auch andere profitieren sobald das releast ist.

Ich warte aktuell noch auf Feedback zu einem anderen Fix, aber danach werde ich die Version 2.2.2 mit allen Fixes releasen.
